discovery-provider/src/tasks/index.py

```diff
@@ -353,6 +372,128 @@ def save_and_get_skip_tx_hash(session, redis):
     return None
 
 
+def group_transaction_by_type(tx_type_to_grouped_lists_map, tx, tx_receipt):
+    tx_target_contract_address = tx["to"]
+    for tx_type, tx_list in tx_type_to_grouped_lists_map.items():
+        tx_is_type = tx_target_contract_address == get_contract_addresses()[tx_type]
+        if tx_is_type:
+            contract_name = helpers.snake_to_pascal(tx_type)
+            logger.info(
+                f"index.py | {contract_name} contract addr: {tx_target_contract_address}"
+                f" tx from block - {tx}, receipt - {tx_receipt}"
+            )
+            tx_list.append(tx_receipt)
+            break
+
+
+def add_indexed_block_to_db(db_session, block):
+    web3 = update_task.web3
+    current_block_query = db_session.query(Block).filter_by(is_current=True)
+
+    # Without this check we may end up duplicating an insert operation
+    block_model = Block(
+        blockhash=web3.toHex(block.hash),
+        parenthash=web3.toHex(block.parentHash),
+        number=block.number,
+        is_current=True,
+    )
+
+    # Update blocks table after
+    assert current_block_query.count() == 1, "Expected single row marked as current"
+
+    previous_block = current_block_query.first()
+    previous_block.is_current = False
+    db_session.add(block_model)
+
+
+def add_indexed_block_to_redis(block, redis):
+    redis.set(most_recent_indexed_block_redis_key, block.number)
+    redis.set(most_recent_indexed_block_hash_redis_key, block.hash.hex())
+
+
+def bulk_process_state_changes(
+    main_indexing_task,
+    session,
+    ipfs_metadata,
+    blacklisted_cids,
+    tx_type_to_grouped_lists_map,
+    block,
+    redis,
+):
+    block_number, block_hash, block_timestamp = itemgetter(
+        "number", "hash", "timestamp"
+    )(block)
+    TX_TYPE_TO_HANDLER_MAP = {
+        USER_FACTORY: user_state_update,
+        TRACK_FACTORY: track_state_update,
+        SOCIAL_FEATURE_FACTORY: social_feature_state_update,
+        PLAYLIST_FACTORY: playlist_state_update,
+        USER_LIBRARY_FACTORY: user_library_state_update,
+        USER_REPLICA_SET_MANAGER: user_replica_set_state_update,
+    }
+    track_metadata_affecting_contracts = [
+        USER_FACTORY,
+        TRACK_FACTORY,
+    ]
+    changed_entity_ids_map = {
+        USER_FACTORY: [],
+        TRACK_FACTORY: [],
+        PLAYLIST_FACTORY: [],
+        USER_REPLICA_SET_MANAGER: [],
+    }
+
+    for tx_type, bulk_processor in TX_TYPE_TO_HANDLER_MAP.items():
+        txs_to_process = tx_type_to_grouped_lists_map[tx_type]
+        tx_processing_args = [
```

Comment:
this part isn't super clean, but unfortunately not all the handlers have the same function signatures. i standardized the return values and can also standardize the params but that would require passing in ipfs_metadata, blacklisted_cids, etc where they aren't needed. thoughts?

Reply:
After some discussion w Dheeraj, I tried both approaches and standardizing the fn sigs for tx processing functions seemed cleaner to me. Happy to change if the team disagrees.

Reply:
I'll do this in a fast follow for this PR where I add integration test coverage for social features + user library state changes, and refactor all the update handlers. To keep this PR review manageable I'm going to limit this one (see https://github.com/AudiusProject/audius-protocol/pull/2254 for latest state) to refactor of the main indexing flow only.